WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/FIle List/ProductDownloadListViewController.swift

```diff
@@ -142,21 +144,73 @@ extension ProductDownloadListViewController {
 // MARK: Action - Add/Edit Product Downloadable File Settings
 //
 extension ProductDownloadListViewController {
-    func addEditDownloadableFile(indexPath: IndexPath) {
+    func addEditDownloadableFile(indexPath: IndexPath, formType: ProductDownloadFileViewController.FormType) {
+        let viewController = ProductDownloadFileViewController(product: product,
+                                                               downloadFileIndex: indexPath.row,
+                                                               formType: formType) { [weak self]
+            (fileName, fileURL, fileID, hasUnsavedChanges) in
+            self?.onAddEditDownloadableFileCompletion(fileName: fileName,
+                                                      fileURL: fileURL,
+                                                      fileID: fileID,
+                                                      hasUnsavedChanges: hasUnsavedChanges,
+                                                      indexPath: indexPath,
+                                                      formType: formType)
+        }
+        navigationController?.pushViewController(viewController, animated: true)
 
     }
 
     func onAddEditDownloadableFileCompletion(fileName: String?,
                                              fileURL: String?,
                                              fileID: String?,
-                                             hasUnsavedChanges: Bool) {
+                                             hasUnsavedChanges: Bool,
+                                             indexPath: IndexPath,
+                                             formType: ProductDownloadFileViewController.FormType) {
+        defer {
+            navigationController?.popViewController(animated: true)
+        }
+
+        guard hasUnsavedChanges else {
+            return
+        }
+
+        switch formType {
+        case .add:
+            viewModel.append(ProductDownloadDragAndDrop(downloadableFile: ProductDownload(downloadID: fileID ?? "",
+                                                                                          name: fileName ?? "",
+                                                                                          fileURL: fileURL ?? "")))
+        case .edit:
+            viewModel.update(at: indexPath.row,
+                             element: (ProductDownloadDragAndDrop(downloadableFile: ProductDownload(downloadID: fileID ?? "",
+                                                                                                    name: fileName ?? "",
+                                                                                                    fileURL: fileURL ?? ""))))
+        }
+        viewModel.completeUpdating(onCompletion: onCompletion)
```

Comment:
on `completeUpdating` there is a `TODO` that maybe should be handled in this PR.

Reply:
Yes, handled.